rfcs/text/design-tokens.md

```diff
@@ -0,0 +1,154 @@
+- Start Date: (fill me in with today's date, 2021-06-14)
+
+
+- [Summary](#summary)
+- [Basic example](#basic-example)
+- [Motivation](#motivation)
+- [Detailed design](#detailed-design)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+    - [Design tokens via styled-components and CSS variables](#design-tokens-via-styled-components-and-css-variables)
+    - [Tag function for generating token](#tag-function-for-generating-token)
+- [Adoption strategy](#adoption-strategy)
+- [How we teach this](#how-we-teach-this)
+
+# Summary
+
+Implementation of the design tokens with theme support and backwards compatibility.
+
+# Basic example
+
+Design tokens will be kept in `@sage/design-tokens` library. It will be updated every time tokens will change. They will
+be imported from this library and used as the object.
+
+```js
+const Container = styled.div`
+  background: ${props.theme.sagedsColorsBase400};
+  color: ${props.theme.sagedsColorsTextWhite};
+`;
+```
+
+Then, themes would be handled by ThemeProvider.
+
+Themes would be created from existing ones by extending and renaming them, so that current *mint* would be named *sbiz*
+.
+
+```js
+import designTokens from "@sage/design-tokens/js/sbiz/es6-obj";
+import { mergeWithBase } from "../base";
+import configureTheme from "./mint-theme.config";
+
+export default {
+  ...mergeWithBase(configureTheme), // for backwards compability. Eventually should be removed.
+  ...designTokens,
+  name: "sbiz",
+};
+```
+
+# Motivation
+
+Design tokens are basic atomic part of design system. Its purpose is to improve handover process between designers and
+developers. This will result in more consistent ui and will help using colors more
+
+Design tokens are basically key-value pairs named and organized the same way regardless of the platform. This way,
+designers and developers can communicate using key names, and not specific values.
+
+Design tokens can contain values like lengths in different units, colors, font properties, transition definitions and
+others.
+
+Naming of the tokens is consistent across all platforms and provides predictable structure.
+
+To put it in simple words: When developer creates design for component, everything what you, as a developer, care about
+is which token has been used.
+
+# Detailed design
```

Comment:
Please can you work out the increase in bundle size and the impact to our consumers. The color codes are generated using functions at the moment. We can't remove the colors at run time because of the `color` prop, although we could remove this functionality if there is a suitable alternative.

Reply:
It relates to the comment I left on L138
https://github.com/Sage/carbon/pull/4163#discussion_r653431651

We couldn't tree shake the unused colors because the consumer can specify the color at runtime so we'd need to include 201 variants of each color (base, + 100 shade, + 100 tint).

Based on your previous reply, maybe this will no longer be supported?

What is the plan for the `color` & `bg` props, where the consumer wants to override the color. Would this accept the colors in the theme e.g. `colorsSlate100`?

Reply:
This would be the big change. Introduction of the Deisgn System will require consumers to adapt to the new rules. If consumer wants to be well aligned with DS, then he has to use theme tokens. 

If he needs a custom color, then he can use generic token (if eventually they will be included), or use custom color and provide it by hex, however then, the responsibility of not aligning to the Design System is on his side.

Reply:
Please can you document this in the RFC as a breaking change and explain how users should upgrade.
e.g. I'm using <Component color="slateShade20" /> how can I upgrade.

What was the outcome of the bundle size? I suspect if we're not supporting every shade and tint for each colour it will not be a problem (i.e. no more `slateShade20`, `slateShade21`, `slateShade22`).

Reply:
dist folder size of the Design Tokens Library pacakge is about 600 KB, but since It will use single theme from DTL, it will use only about 20 KB, and should tree-shae the rest of it.

I'll update RFC with and bundle size.

Reply:
It's not clear how tree shaking will reduce the size of the CSS file? Isn't it true that we'll need to include all CSS properties because we don't know which ones are used at runtime?

Looks like my previous comment was truncated.

> e.g. I'm using how can I upgrade.

should have been

> e.g. I'm using `<Component color={slateShade20}  bg={slateShade21}/>`how can I upgrade.

Reply:
During the time of implementation, we're going to use js variables, that will be translated into css variables and provided with `GlobalStyle`.
In package there will be available different sheets - scss, css, js (es6 and commonjs).
Also, there will be available different themes.
Overall size of package is about 600KB at the time.
Most probably, Carbon will use only one js file (es6 or commonjs) from the whole package.
So what I meant by saying "treeshaking" is that so all the others wont be bundled into final Carbon build.

> e.g. I'm using <Component color={slateShade20} bg={slateShade21}/>how can I upgrade.

You shouldn't be overwriting values since it breaks component alignment with the DS documentation.
All properties that need to be set, should be set inside component.
However if you really need to overwrite any property, you can do it by redefining css variable locally, for given component. e.g. in class definition.

Reply:
To be clear, DS team will remove the existing `color` and `bg` props from all components?

Reply:
After consideration, as we agreed on, `color` and `bg` props shouldn't be removed.
Also, styling inner components cannot be done by referring to inner elements by css path.

So the way it should work is to pass color using design token name:
```
<CarbonComponent color="sageColorAction500" />
```
so the `color` prop should accept token name and use its value or preserve existing logic for retrieving value.